iree/compiler/Conversion/LinalgToSPIRV/FoldGPUProcessorIDUses.cpp

```diff
@@ -0,0 +1,268 @@
+  LogicalResult matchAndRewrite(AffineMinOp minOp,
+                                PatternRewriter &rewriter) const override {
+    LLVM_DEBUG(llvm::dbgs() << "inspecting " << minOp << "\n");
+    MLIRContext *context = minOp.getContext();
+    auto dimensions = minOp.getDimOperands();
+    auto symbols = minOp.getSymbolOperands();
+
+    // We expect the affine.min op to only have one symbol operand.
+    if (!llvm::hasSingleElement(symbols) || !dimensions.empty()) {
+      return rewriter.notifyMatchFailure(
+          minOp, "expected to only have one symbol operand");
+    }
+
+    // And the symbol operand should come from a GPU processor ID.
+    Operation *symbolOp = symbols.front().getDefiningOp();
+    auto symbol0 = getAffineSymbolExpr(0, context).cast<AffineSymbolExpr>();
+
+    Optional<int64_t> ub;
+    if (auto blockIDOp = dyn_cast<gpu::BlockIdOp>(symbolOp)) {
+      ub = getProcessorIDUpperBound(blockIDOp);
+    } else if (auto threadIDOp = dyn_cast<gpu::ThreadIdOp>(symbolOp)) {
+      ub = getProcessorIDUpperBound(threadIDOp);
+    }
+    if (!ub) {
+      return rewriter.notifyMatchFailure(
+          minOp, "failed to query processor ID upper bound");
+    }
+    LLVM_DEBUG(llvm::dbgs() << "processor ID '" << *symbolOp
+                            << "' upper bound: " << *ub << "\n");
+
+    // Look at each result expression. For expressions that are functions of
+    // the input symbol, try to simplify it. We do this by replacing the
+    // symbol with its lower and upper bound. This requires the result
+    // expression to be a linear function of the input symbol.
+    SmallVector<AffineExpr, 4> results;
+    // The indices into `results` where the corresponding AffineExpr is a
+    // constant from the original map. We need to keep track of this so later we
+    // can probe whether the constant is the min.
+    SmallVector<unsigned, 4> cstIndices;
+    for (auto result : minOp.getAffineMap().getResults()) {
+      if (auto cstResult = result.dyn_cast<AffineConstantExpr>()) {
+        results.push_back(cstResult);
+        cstIndices.push_back(results.size() - 1);
+      } else if (isLinearExpr(result)) {
+        results.push_back(simplifyAffineExpr(
+            replaceSymbolWithValue(result, symbol0, 0), 0, 1));
+        results.push_back(simplifyAffineExpr(
+            replaceSymbolWithValue(result, symbol0, *ub - 1), 0, 1));
+        LLVM_DEBUG({
+          auto map = AffineMap::get(0, 1, results, context);
+          llvm::dbgs() << "map after substituting with processor ID bounds: "
+                       << map << "\n";
+        });
+      } else {
+        // We cannot handle such cases. Just bail out on matching the pattern.
+        return rewriter.notifyMatchFailure(
+            minOp, "expected to have a linear function of the symbol");
+      }
+    }
+
+    auto isNegative = [](AffineExpr e) {
+      if (auto cst = e.dyn_cast<AffineConstantExpr>())
+        return cst.getValue() < 0;
+      return true;
```

Comment:
Returning `true` by default for `isNegative` is hard to reason about. It might be easier to reason about this if is renamed?